src/main/target/REVONANO/target.h

```diff
@@ -79,7 +79,11 @@
 
 #define USE_I2C
 #define USE_I2C_DEVICE_1        // UART1/FlexiPort (PB6,PB7)
-#define I2C_DEVICE (I2CDEV_1)
```

Comment:
This should be restored as well (`I2C_DEVICE` is used to set the I2C device for the dashboard).

Reply:
Actually, on a closer look, the I2C device for the dashboard can just as well be set by defining `DASHBOARD_I2C_INSTANCE`, just like all other I2C drivers have custom defines for the I2C device. So I guess the only justification for defining `I2C_DEVICE` is if the bus is broken out and can potentially be used for a number of devices. Or am I missing something here @jflyper?

Reply:
`I2C_DEVICE` provides a default for devices, and most devices also have their own `xxx_I2C_DEVICE` which actually defaults to `I2C_DEVICE`.

> the only justification for defining I2C_DEVICE is if the bus is broken out and can potentially be used for a number of devices.

"The only" is too strong, but yes, that is one way of using it.

Reply:
If I change `I2C_DEVICE` to `I2CDEV_1` then the Baro isn't found because the ms5611Detect() fails. This is because the first I2C access on I2C bus 3 fails. If `I2C_DEVICE` is defined as `I2CDEV_3` then the resulting probe of the MPU on I2C bus 3 fixes this and the subsequent call to ms5611Detect() works. I've therefore added a dummy read of the `CMD_PROM_RD` register in ms5611Detect() and thereafter the probe works.

Reply:
> If I2C_DEVICE is defined as I2CDEV_3 then the resulting probe of the MPU on I2C bus 3 fixes this

I don't like the code that relies on coincidence...

> If I change I2C_DEVICE to I2CDEV_1 then the Baro isn't found because the ms5611Detect() fails. This is because the first I2C access on I2C bus 3 fails.

We shouldn't take pragmatic approach to fix this.

We must find out if the failure to detect MS5611 has been around for a long time, whether is local to REVONANO, local to F411CE, local to I2C3, and such, and then implement a fix.

Reply:
@jflyper On the first access the `i2c_er_handler()` ISR of the I2C bus driver gets called. The status register `SR1` has `I2C_SR1_AF` set indicating that no device acknowledged the address sent in the I2C transaction. This happens after a prior call to `i2c_ev_handler()` in turn calling `I2C_Send7bitAddress()` so the address has been sent to the bus. In other words, this looks like a well formed access to the Baro, that it simply fails to acknowledge.

The data sheet for the MS5611 indicates that the device will by busy (which means it won't respond to any I2C accesses) during a conversion. The data sheet is a bit thin to be honest. It's worth noting that OpenPilot code does not consider ACK failures from this device to be errors and carries on regardless.

Reply:
Testing on a Revo the dummy read isn't needed, but it's worth noting that that has both the MS5611 Baro and HMC5883 Mag on the same I2C bus, so perhaps it's enough to have any I2C cycle to enable access, not specifically one to the MS5611.

Reply:
Removing the mag from the build makes no difference; the baro still works without a dummy access.

Reply:
The Nano has a 24C512 EEPROM on the same bus, which also fails to acknowledge if it's busy. Interestingly the Revo has 2K2 pull-ups on the I2C bus, but the Nano has 10K. Not sure if that makes any difference.

Reply:
As the EEPROM isn't used on the RevoNano it would seem, it really shouldn't have any impact. I've tried disabling I2C overclocking, and enabling pull-ups on the line, but to no benefit. It still errors the first access to the Baro.

Reply:
@jflyper Accessing any I2C address unsticks the bus. Accessing the I2C EEPROM before the Baro (at address 0x50 on I2C bus 3) stops the issue without an error, so the dummy read I added doesn't need to be to the Baro, and won't fail if it's to a legit device.

It therefore seems that it's just the Baro that's upset and only then for the first access. I noticed that i2cUnstick() is called when the bus is initialised. If I don't call that the problem goes away which suggest that it's this that upsets the Baro. I've rewritten this routine to emit a valid I2C transaction, and now the Baro works every time. 😃